File: LinkedList/40reverseLinkedList/reverseLinkedList.java

//  * Definition for singly-linked list.
// class ListNode {
//     int val;
//     ListNode next;

//     ListNode() {
//     }

//     ListNode(int val) {
//         this.val = val;
//     }

//     ListNode(int val, ListNode next) {
//         this.val = val;
//         this.next = next;
//     }
// }

public class reverseLinkedList {
    static class ListNode {
        private int val;
        private ListNode next;

        public ListNode(int val) {
            this.val = val;
        }

        public void setNext(ListNode node) {
            this.next = node;
        }

        public ListNode getNext() {
            return this.next;
        }
    }

    public static ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode next_node = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next_node;
        }
        return prev;
    }

    public static void printLinkedList(ListNode node) {
        ListNode top = node;
        while (top != null) {
            System.out.print(top.val + " ");
            top = top.next;
        }
    }

    public static void main(String[] args) {
        ListNode n5 = new ListNode(5);
        ListNode n4 = new ListNode(4);
        ListNode n3 = new ListNode(3);
        ListNode n2 = new ListNode(2);
        ListNode n1 = new ListNode(1);
        n1.setNext(n2);
        n2.setNext(n3);
        n3.setNext(n4);
        n4.setNext(n5);
        ListNode reversed = reverseList(n1);
        System.out.println();
        printLinkedList(reversed);
    }
}

File: LinkedList/40reverseLinkedList/reverseLinkedList.py
# Reverse Linked List - leetcode 206
# 1->2->3->4->5->NULL ; input -> ouput: 5->4->3->2->1->NULL
class ListNode:
    def __init__(self, val = 0, nexty=None):
        self.val = val
        self.nexty = nexty

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev, curr = None, head
# T: O(n), M(1) : iterative
# T: O(n), M(n) : recursive
        while curr:
            nxt = curr.nexty
            curr.nexty = prev
            prev = curr
            curr = nxt
        
        return prev

    def recurse(self, head: ListNode) -> ListNode:
        if not head:
            return None
        newHead = head
        if head.nexty:
            newHead = self.recurse(head.nexty)
            head.nexty.nexty = head
        head.nexty = None

        return newHead